lib/Yare/App/Services/Rebalancing.hs

```diff
@@ -0,0 +1,279 @@
+import Yare.Util.Tx.Construction (minAdaValue, mkCardanoApiUtxo, witnessUtxoEntry)
+import Yare.Utxo (Entry (..), Utxo, spendableEntries)
+import Yare.Utxo qualified as Utxo
+
+service
+  ∷ ∀ era state env
+   . ( [Addresses, Submitter.Q, NetworkInfo era, StorageMgr IO state] ∈∈ env
+     , '[Utxo] ∈∈ state
+     )
+  ⇒ env
+  → IO TxId
+service env = do
+  let NetworkInfo {currentEra} = look @(NetworkInfo era) env
+  let submitQueue ∷ Submitter.Q = look env
+  let storageManager ∷ StorageMgr IO state = look env
+  setStorageMode storageManager Durable
+  overDefaultStorage storageManager rebalance' \case
+    Left err → do
+      putTextLn "Error while rebalancing:"
+      putTextLn . toStrict $ pShow err
+      throwIO err
+    Right tx → do
+      putTextLn "Submitting the transaction:"
+      putTextLn . toStrict $ pShow tx
+      let txInMode = TxInMode (convert currentEra) tx
+      getTxId (getTxBody tx)
+        <$ Submitter.submit submitQueue txInMode
+ where
+  rebalance' ∷ state → (state, Either Error (Tx era))
+  rebalance' s =
+    case runExcept (runStateT (rebalance env) s) of
+      Left e → (s, Left e)
+      Right (r, s') → (s', Right r)
+
+rebalance
+  ∷ ∀ state env era
+   . ( '[NetworkInfo era, Addresses] ∈∈ env
+     , '[Utxo] ∈∈ state
+     )
+  ⇒ env
+  → StateT state (Except Error) (Tx era)
+rebalance env = do
+  let
+    addresses = look @Addresses env
+    rebalancingAddresses = externalAddresses addresses
+    network ∷ NetworkInfo era = look env
+    era ∷ ConwayEraOnwards era = currentEra network
+    epoch = epochInfo network
+    protocolParams = protocolParameters network
+    shelleyBasedEra ∷ ShelleyBasedEra era = convert era
+
+    changeAddress ∷ AddressInEra era =
+      fromShelleyAddrIsSbe shelleyBasedEra . ledgerAddress $
+        Address.useForChange addresses
+
+    minUtxoAda ∷ Lovelace =
+      calculateMinimumUTxO shelleyBasedEra dummyTxOut protocolParams
+     where
+      dummyTxOut ∷ TxOut CtxTx era =
+        TxOut changeAddress dummyBalance TxOutDatumNone ReferenceScriptNone
+      dummyBalance ∷ TxOutValue era =
+        lovelaceToTxOutValue shelleyBasedEra 10_000_000
+
+    expectedTotalFee ∷ Lovelace = 500_000
+    expectedTotalCollateral ∷ Lovelace = 600_000
+
+  utxoEntryForFee ∷ Utxo.Entry ←
+    usingMonadState (Utxo.useInputFee addresses (minUtxoAda + expectedTotalFee))
+      >>= maybe (throwError (RebalancingTxError NoFeeInputs)) pure
+
+  utxoEntryForCollateral ∷ Utxo.Entry ←
+    usingMonadState (Utxo.useInputCollateral addresses (minUtxoAda + expectedTotalCollateral))
+      >>= maybe (throwError (RebalancingTxError NoCollateralInputs)) pure
+
+  totalLovelaceBalance ∷ Lovelace ←
+    usingMonadState (calculateTotalBalance rebalancingAddresses)
+      >>= maybe (throwError CalculateTotalBalanceError) pure
+
+  rebalanceEntries ∷ [Utxo.Entry] ←
+    usingMonadState (useInputsForRebalancing addresses)
+      >>= maybe (impossible "useInputsForRebalancing resulted in Nothing") pure
```

Comment:
```suggestion
      >>= maybe (throwError (RebalancingTxError NoInputsForRebalancing)) pure
```

Reply:
Or why do you think this is impossible? If so, then maybe types should express such impossibility, e.g. don't use Maybe